### ISSUE TITLE
database is locked (OperationalError)

### ISSUE THREAD

**Issue by ppetoumenos** [ChrisCummins/clgen#41]
Both when running make test and running clgen model.json sampler.json, I get an error like this:

```
spawning 32 worker threads to process 250 files ...
fatal: database is locked (OperationalError)

```
If I run clgen again, it uses whatever was written in the kernel.db before the error, does the training, but gets stuck. After that nothing happens, luajit is in Zombie mode and I cannot kill it, and rerunning clgen does absolutely nothing.

```
creating DNN corpus /home/ppetou/.cache/clgen/0.0.24/6b4248154caa140665461619e4d1292eeee1b225/corpus.txt.tmp ...
writing file /home/ppetou/.cache/clgen/0.0.24/6b4248154caa140665461619e4d1292eeee1b225/corpus.txt.tmp
Total vocabulary size: 77
Total tokens in file: 86418
  Training size: 69136
  Val size: 8641
  Test size: 8641
Using dtype  <type 'numpy.uint8'>

Running with CUDA on GPU 0	
Epoch 1.37 / 10, i = 10 / 270, loss = 2.770734	
Epoch 1.74 / 10, i = 20 / 270, loss = 2.675026	
Epoch 2.11 / 10, i = 30 / 270, loss = 2.582568	
Epoch 2.48 / 10, i = 40 / 270, loss = 2.384112	
Epoch 2.85 / 10, i = 50 / 270, loss = 2.138002	
Epoch 3.22 / 10, i = 60 / 270, loss = 1.960270	
Epoch 3.59 / 10, i = 70 / 270, loss = 1.692933	
Epoch 3.96 / 10, i = 80 / 270, loss = 1.475670	
Epoch 4.33 / 10, i = 90 / 270, loss = 1.265503	
Epoch 4.70 / 10, i = 100 / 270, loss = 1.092704	
val_loss = 	3.3303287029266	
Epoch 5.07 / 10, i = 110 / 270, loss = 1.027073	
Epoch 5.44 / 10, i = 120 / 270, loss = 0.892996	
Epoch 5.81 / 10, i = 130 / 270, loss = 0.775659	
Epoch 6.19 / 10, i = 140 / 270, loss = 0.827018	
Epoch 6.56 / 10, i = 150 / 270, loss = 0.640372	
Epoch 6.93 / 10, i = 160 / 270, loss = 0.618316	
Epoch 7.30 / 10, i = 170 / 270, loss = 0.509315	
Epoch 7.67 / 10, i = 180 / 270, loss = 0.459649	
Epoch 8.04 / 10, i = 190 / 270, loss = 0.741571	
Epoch 8.41 / 10, i = 200 / 270, loss = 0.385434	
val_loss = 	2.9984436035156	
Epoch 8.78 / 10, i = 210 / 270, loss = 0.369498	
Epoch 9.15 / 10, i = 220 / 270, loss = 0.391619	
Epoch 9.52 / 10, i = 230 / 270, loss = 0.327276	
Epoch 9.89 / 10, i = 240 / 270, loss = 0.280787	
Epoch 10.26 / 10, i = 250 / 270, loss = 0.299564	
Epoch 10.63 / 10, i = 260 / 270, loss = 0.306531	
Epoch 11.00 / 10, i = 270 / 270, loss = 0.440610	
val_loss = 	2.8170424302419	
```

**Comment by ChrisCummins** [comment]
Thanks. So the clgen process lifetime looks something like this:

<img width="850" alt="screen shot 2016-11-03 at 14 01 26" src="https://cloud.githubusercontent.com/assets/1636663/19968847/2ccf270c-a1ce-11e6-9eed-30a9374f2c39.png">

It runs the luaj interpreter to do sampling, then forks itself 32 times to do the preprocessing. When you say the zombie luajit, do you mean that you can't C-c out of clgen during the sampling bit? Or that you can C-c to get back to the command line, but there's some process knocking around in the background? If it's the former, I've had that problem too and it's super annoying but I'm not sure if I can fix it.

The OperationalError is from the 32 python processes competing for the same SQL database. I can fix that

**Comment by ppetoumenos** [comment]
The problem comes after preprocessing during the training. After the last iteration nothing happens. I can C-c, terminate the script, and get back to the command line, but the luajit process is not terminated and cannot be terminated in any way except for rebooting the machine.

This on its own might not be a problem but I assume this is the reason that running again clgen does nothing.

**Comment by ppetoumenos** [comment]
the problem persists even with 8 worker threads

**Comment by ppetoumenos** [comment]
With 2 worker threads it runs fine. Perhaps we should reduce it?

The zombie luajit thing was because one of its threads was stuck inside the GPU driver. I upgraded the driver and the problem went away.

**Comment by ChrisCummins** [comment]
Thanks that's good to know! What driver version was it running, and what did you update to?

For now, set `num_workers=1` [here](https://github.com/ChrisCummins/clgen/blob/master/clgen/preprocess.py#L477) to fix the database crash, though I'll push a more permanent fix soon.

**Comment by ppetoumenos** [comment]
was 352.39, now it's 367.57

**Comment by ChrisCummins** [comment]
Thanks. Doesn't seem that old. The only thing I see in the torch docs is CUDA >= 6.5.

I may add a `--cpu` flag to help debug GPU issues.

**Comment by ChrisCummins** [comment]
Database locking fixed in [0.0.31](https://github.com/ChrisCummins/clgen/releases/tag/0.0.31) 🎉 

```
$ curl -s https://raw.githubusercontent.com/ChrisCummins/clgen/0.0.30/install-cuda.sh | bash
```
